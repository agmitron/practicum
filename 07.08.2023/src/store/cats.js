import {
  action,
  makeObservable,
  observable,
  computed,
  autorun,
  runInAction,
} from "mobx";

class Cats {
  data = [];

  constructor() {
    makeObservable(this, {
      data: observable,
      addCat: action,
      deleteCat: action,
      count: computed,
      loadInitialKitties: action
    });

    autorun(() => this.log());
    runInAction(() => this.loadInitialKitties());
  }

  addCat(url) {
    const id = this.data.length + 1;
    this.data.push({ url, id });
  }

  deleteCat(id) {
    this.data = this.data.filter((cat) => cat.id !== id);
  }

  get count() {
    return this.data.length;
  }

  log() {
    console.log(`Котики: ${this.data.map((cat) => cat.url).join(",")}`);
  }

  async loadInitialKitties() {
    const response = await fetch(`https://cataas.com/api/cats?skip=0&limit=30`);
    const data = await response.json();

    this.data = data.map((cat, id) => {
      const text = cat.owner + cat.tags.join(",") + id;
      const url = `https://cataas.com/cat/says/${text}`;

      return { url, id };
    });
  }
}

const cats = new Cats();

export default cats;
