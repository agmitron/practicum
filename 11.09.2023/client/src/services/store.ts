import { configureStore } from '@reduxjs/toolkit'
import {
  TypedUseSelectorHook,
  useDispatch as dispatchHook,
  useSelector as selectorHook
} from 'react-redux';

import rootReducer from './reducers';

import { socketMiddleware } from './middleware/socket-middleware';

import { 
  connect as LiveTableWsConnect, 
  disconnect as LiveTableWsDisconnect,
  wsConnecting as LiveTableWsConnecting,
  wsOpen as LiveTableWsOpen,
  wsClose as LiveTableWsClose,
  wsMessage as LiveTableWsNessage,
  wsError as LiveTableWsError 
} from "./reducers/live-table/actions";

import { TLiveTableActions } from './reducers/live-table/actions';
import {ThunkAction} from "redux-thunk";

const wsActions = {
  wsConnect: LiveTableWsConnect,
  wsDisconnect: LiveTableWsDisconnect,
  wsConnecting: LiveTableWsConnecting,
  onOpen: LiveTableWsOpen,
  onClose: LiveTableWsClose,
  onError: LiveTableWsError,
  onMessage: LiveTableWsNessage,
};

const liveTableMiddleware = socketMiddleware(wsActions);

export type RootState = ReturnType<typeof rootReducer>

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) => {
    return getDefaultMiddleware().concat(liveTableMiddleware)
  }
})
/*
export type AppDispatch = typeof store.dispatch
export const useDispatch = () => dispatchHook<AppDispatch>()
*/

export type AppActions = TLiveTableActions;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  AppActions
>

type Dispatch = <TReturnType>(action: AppActions | AppThunk) => TReturnType;
export const useDispatch = () => dispatchHook<Dispatch>();
export const useSelector: TypedUseSelectorHook<RootState> = selectorHook;

