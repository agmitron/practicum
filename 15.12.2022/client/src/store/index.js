import { createStore, combineReducers } from 'redux'
import shop from './shop/reducer'
import user from './user/reducer'

const reducers = combineReducers({
  shop,
  user
})

export default createStore(reducers)