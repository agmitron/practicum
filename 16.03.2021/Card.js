class Component {
    render() {}
    _setEventListeners() {}
}

class Card extends Component {
    constructor(templateSelector, data, imagePopup) {
        super()
        this.data = data
        this.$template = document.querySelector(templateSelector)
        this.imagePopup = imagePopup
    }

    _setEventListeners(handlers) {
        handlers.forEach(h => {
            h.el.addEventListener(h.listener, h.handler)
        })
    }

    _remove($parent, $el) {
        const really = confirm('Вы действительно хотите удалить карточку?')

        if (really) {
            $parent.removeChild($el)
        }
    }

    _like($like) {
        $like.classList.toggle('button_type_like_active_true')
        this.data.liked = !this.data.liked
    }

    _openImage() { 
        this.imagePopup.open(this.data.imgUrl, this.data.title)
    }

    render(parentSelector = 'body') {
        const { id, title, liked, imgUrl } = this.data

        const $el = this.$template.content.cloneNode(true).querySelector('.card')
        const $image = $el.querySelector('.card__image')
        const $title = $el.querySelector('.card__heading')
        const $like = $el.querySelector('.button_type_like')
        const $delete = $el.querySelector('.button_type_delete')

        $image.setAttribute('src', imgUrl)
        $title.textContent = title

        if (liked) {
            $like.classList.add('button_type_like_active_true')
        }

        const $parent = document.querySelector(parentSelector)
        $parent.insertAdjacentElement('beforeend', $el)

        this._setEventListeners([
            {
                el: $like, 
                listener: 'click', 
                handler: () => this._like($like)
            },
            {
                el: $delete, 
                listener: 'click',
                handler: () => this._remove($parent, $el)
            },
            {
                el: $image, 
                listener: 'click',
                handler: () => this._openImage()
            }
        ])
    }
}

class ImagePopup extends Component {
    constructor(selector) {
        super()
        this.$el = document.querySelector(selector)
        this._setEventListeners()
    }

    close() {
        this.$el.hidden = true
    }

    open(imgUrl, title) {
        this.$el.querySelector('img').setAttribute('src', imgUrl)
        this.$el.querySelector('h3').textContent = title
        this.$el.hidden = false 
    }

    _setEventListeners() {
        this.$el.addEventListener('click', () => this.close())
    }
}

const json = [
    {
        id: 0,
        title: 'Карачаевск',
        liked: false,
        imgUrl: 'https://upload.wikimedia.org/wikipedia/commons/c/c7/Mount_Elbrus_May_2008.jpg'
    },
    {
        id: 1,
        title: 'Эльбрус',
        liked: true,
        imgUrl: 'https://strahu-net.com/upload/sprint.editor/eea/img_1511791232_673_poxod_na_elbrus_v_mae_2016_1.jpg'
    },
    {
        id: 2,
        title: 'Домбай',
        liked: false,
        imgUrl: 'https://mwtravel.ru/wp-content/uploads/2018/10/Jelbrus-ves-640x360.jpg'
    },
    {
        id: 3,
        title: 'Португалия',
        liked: false,
        imgUrl: 'https://mwtravel.ru/wp-content/uploads/2018/10/Jelbrus-ves-640x360.jpg'
    }
]

const CARD_TEMPLATE_SELECTOR = '.card-template'

const imagePopup = new ImagePopup('.image-popup')

json.forEach(data => {
    const card = new Card(CARD_TEMPLATE_SELECTOR, data, imagePopup)
    card.render()
})