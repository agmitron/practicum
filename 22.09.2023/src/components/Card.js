import React from "react";
import PropTypes from "prop-types";
import { cardPropTypes } from "../utils/prop-types";
import { useSelector, useDispatch } from "react-redux";
import { getCurrentUser } from "../store/current-user/selectors";
import { changeLikeCardStatus } from "../store/cards/actions";

function Card({ card, onImageClick, onDelete }) {
  const dispatch = useDispatch();
  const currentUser = useSelector(getCurrentUser);

  function handleImageClick() {
    onImageClick(card);
  }

  const isLiked = card.likes.some((i) => i._id === currentUser._id);
  const cardLikeButtonClassName = `card__like-button ${
    isLiked && "card__like-button_is-active"
  }`;

  const isOwn = card.owner._id === currentUser._id;
  const cardDeleteButtonClassName = `card__delete-button ${
    isOwn ? "card__delete-button_visible" : "card__delete-button_hidden"
  }`;

  function handleLikeClick() {
    dispatch(changeLikeCardStatus(card._id, !isLiked));
  }

  function handleDeleteClick() {
    onDelete(card);
  }

  return (
    <li className='places__item card'>
      <div
        className='card__image'
        style={{ backgroundImage: `url(${card.link})` }}
        onClick={handleImageClick}
      ></div>
      <button
        type='button'
        className={cardDeleteButtonClassName}
        onClick={handleDeleteClick}
      ></button>
      <div className='card__description'>
        <h2 className='card__title'>{card.name}</h2>
        <div className='card__likes'>
          <button
            type='button'
            className={cardLikeButtonClassName}
            onClick={handleLikeClick}
          ></button>
          <p className='card__like-count'>{card.likes.length}</p>
        </div>
      </div>
    </li>
  );
}

Card.propTypes = {
  card: cardPropTypes,
  onImageClick: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
};

export default Card;
