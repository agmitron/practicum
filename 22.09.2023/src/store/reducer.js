import {combineReducers} from 'redux';

import {name as currentUserStoreName} from './current-user/actions';
import currentUser from './current-user/reducer';
import {name as cardsStoreName} from './cards/actions';
import cards from './cards/reducer';
import {name as authStoreName} from './auth/actions';
import auth from './auth/reducer';

export default combineReducers({
  [currentUserStoreName]: currentUser,
  [cardsStoreName]: cards,
  [authStoreName]: auth,
});
