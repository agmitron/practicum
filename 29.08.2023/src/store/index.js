import { createStore, combineReducers, applyMiddleware } from 'redux'
import createSagaMiddleware from 'redux-saga'

import shop from './shop/reducer'
import user from './user/reducer'
import thunk from 'redux-thunk'
import mySaga from './saga'
const sagaMiddleware = createSagaMiddleware()

const reducers = combineReducers({
  shop,
  user
})

const store = createStore(reducers, applyMiddleware(thunk, sagaMiddleware))
sagaMiddleware.run(mySaga)
export default store