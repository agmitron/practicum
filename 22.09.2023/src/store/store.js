import { createStore, applyMiddleware  } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import reducer from "./reducer";
import mestoApi from "../utils/mesto-api";
import * as authApi from "../utils/auth-api";

const store = createStore(
  reducer,
  composeWithDevTools(  
    applyMiddleware(thunk.withExtraArgument({mestoApi, authApi})),
  )
);

export default store;