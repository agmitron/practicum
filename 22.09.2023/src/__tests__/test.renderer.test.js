import React from 'react';
import { Provider } from 'react-redux';
import TestRenderer from 'react-test-renderer';
import Card from '../components/Card';
import store from "../store/store"

describe('Check Card component', () => {
  let mockCard = {
    "likes": [],
    "_id": "1",
    "name": "Deer",
    "link": "https://images.unsplash.com/",
    "owner":
    {
        "name": "John",
        "about": "Web-developer",
        "avatar": "https://images.unsplash.com/",
        "_id": "1",
        "cohort": "cohort-13"
    },
    "createdAt": "2021-10-24T19:41:51.239Z"
  };

  test('should be equal to snapshot', () => {
    const renderedValue = TestRenderer.create(<Provider store={store}>
      <Card card={mockCard} onImageClick={()=>{}} onDelete={()=>{}}/></Provider>);
    expect(renderedValue).toMatchSnapshot();
  })
})