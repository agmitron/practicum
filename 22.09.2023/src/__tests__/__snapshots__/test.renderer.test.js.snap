// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Card component should be equal to snapshot 1`] = `
<li
  className="places__item card"
>
  <div
    className="card__image"
    onClick={[Function]}
    style={
      Object {
        "backgroundImage": "url(https://images.unsplash.com/)",
      }
    }
  />
  <button
    className="card__delete-button card__delete-button_hidden"
    onClick={[Function]}
    type="button"
  />
  <div
    className="card__description"
  >
    <h2
      className="card__title"
    >
      Deer
    </h2>
    <div
      className="card__likes"
    >
      <button
        className="card__like-button false"
        onClick={[Function]}
        type="button"
      />
      <p
        className="card__like-count"
      >
        0
      </p>
    </div>
  </div>
</li>
`;
